nixos/modules/services/networking/hostapd.nix

```
@@ -1,222 +1,1228 @@
+      ++ singleton ''
+        ssid=${bssCfg.ssid}
+        utf8_ssid=${bool01 bssCfg.utf8Ssid}
 
-  escapedInterface = utils.escapeSystemdPath cfg.interface;
+        logger_syslog=-1
+        logger_syslog_level=${toString bssCfg.logLevel}
+        logger_stdout=-1
+        logger_stdout_level=${toString bssCfg.logLevel}
+        ctrl_interface=/run/hostapd
+        ctrl_interface_group=${bssCfg.group}
 
-  configFile = pkgs.writeText "hostapd.conf" ''
-    interface=${cfg.interface}
-    driver=${cfg.driver}
-    ssid=${cfg.ssid}
-    hw_mode=${cfg.hwMode}
-    channel=${toString cfg.channel}
-    ieee80211n=1
-    ieee80211ac=1
-    ${optionalString (cfg.countryCode != null) "country_code=${cfg.countryCode}"}
-    ${optionalString (cfg.countryCode != null) "ieee80211d=1"}
+        macaddr_acl=${bssCfg.macAcl}''
+      ++ optional (bssCfg.macAllow != [] || bssCfg.macAllowFile != null || bssCfg.authentication.saeAddToMacAllow)
+        "accept_mac_file=/run/hostapd/${bss}.mac.allow"
+      ++ optional (bssCfg.macDeny != [] || bssCfg.macDenyFile != null)
+        "deny_mac_file=/run/hostapd/${bss}.mac.deny"
+      ++ singleton ''
+        # Only allow WPA, disable insecure WEP
+        auth_algs=1
+        ignore_broadcast_ssid=${bssCfg.ignoreBroadcastSsid}
+        # Always enable QoS, which is required for 802.11n and above
+        wmm_enabled=1
+        ap_isolate=${bool01 bssCfg.apIsolate}
 
-    # logging (debug level)
-    logger_syslog=-1
-    logger_syslog_level=${toString cfg.logLevel}
-    logger_stdout=-1
-    logger_stdout_level=${toString cfg.logLevel}
+        # IEEE 802.11i (authentication) related configuration
+        # Encrypt management frames to protect against deauthentication and similar attacks
+        ieee80211w=${bssCfg.managementFrameProtection}''
+      ++ optional (bssCfg.authentication.mode == "none")
+        "wpa=0"
+      ++ optional (bssCfg.authentication.mode == "wpa3-sae") ''
+        wpa=2
+        wpa_key_mgmt=SAE
+        # Derive PWE using both hunting-and-pecking loop and hash-to-element
+        sae_pwe=2
+        # Prevent downgrade attacks by indicating to clients that they should
+        # disable any transition modes from now on.
+        transition_disable=0x01''
+      ++ optional (bssCfg.authentication.mode == "wpa3-sae-transition") ''
+        wpa=2
+        wpa_key_mgmt=WPA-PSK-SHA256 SAE''
+      ++ optional (bssCfg.authentication.mode == "wpa2-sha256") ''
+        wpa=2
+        wpa_key_mgmt=WPA-PSK-SHA256''
+      ++ optional (bssCfg.authentication.mode != "none") ''
+        wpa_pairwise=${pairwiseCiphers}
+        rsn_pairwise=${pairwiseCiphers}''
+      ++ optional (bssCfg.authentication.wpaPassword != null)
+        "wpa_passphrase=${bssCfg.authentication.wpaPassword}"
+      ++ optional (bssCfg.authentication.wpaPskFile != null)
+        "wpa_passphrase=${bssCfg.authentication.wpaPskFile}"
+      ++ (map genSaePasswordEntry bssCfg.authentication.saePasswords)
```

I think you're missing a newline here, in case both wpaPasswords and `wpaPskFile` are null:
```suggestion
      ++ (map genSaePasswordEntry bssCfg.authentication.saePasswords)
      ++ optional (bssCfg.authentication.wpaPasswordFile != null ||
                   bssCfg.authentication.saePasswordsFile != null) ""
```

---

No, you'll get an invalid configuration file. I'm just not sure where it's best to put the newline: either here or in the script that concatenates the files.

---

Could this be an more general issue because the file doesn't end with a newline (because of `concatStringsSep "\n"`)? If so, this should fail in every case and I should probably use `concatStringsMap (x: "${x}\n")`.